src/main/java/com/hierynomus/smbj/Config.java

```diff
@@ -15,36 +15,182 @@
 import com.hierynomus.smbj.auth.Authenticator;
+import com.hierynomus.smbj.auth.NtlmAuthenticator;
+import com.hierynomus.smbj.auth.SpnegoAuthenticator;
+
+public final class Config {
+    private Set<SMB2Dialect> dialects;
+    private List<Factory.Named<Authenticator>> authenticators;
+    private Random random;
+    private UUID clientGuid;
+    private boolean signingRequired;
+    private SecurityProvider securityProvider;
+    private int readBufferSize;
+    private int writeBufferSize;
+    private int transactBufferSize;
+
+    public static Config createDefaultConfig() {
+        Config.Builder b = builder();
+
+        b.setRandomProvider(new SecureRandom());
+        b.setSecurityProvider(new JceSecurityProvider());
+        b.addDialect(SMB2Dialect.SMB_2_1);
+        b.addDialect(SMB2Dialect.SMB_2_0_2);
+        b.setSigningRequired(false);
+
+        // order is important.  The authenticators listed first will be selected
+        b.addAuthenticator(new SpnegoAuthenticator.Factory());
+        b.addAuthenticator(new NtlmAuthenticator.Factory());
+
+        return b.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private Config() {
+        dialects = EnumSet.noneOf(SMB2Dialect.class);
+        authenticators = new ArrayList<>();
+    }
+
+    private Config(Config other) {
+        this();
+        dialects.addAll(other.dialects);
+        authenticators.addAll(other.authenticators);
+        random = other.random;
+        clientGuid = other.clientGuid;
+        signingRequired = other.signingRequired;
+        securityProvider = other.securityProvider;
+        readBufferSize = other.readBufferSize;
+        writeBufferSize = other.writeBufferSize;
+        transactBufferSize = other.transactBufferSize;
+    }
+
+    public Random getRandomProvider() {
+        return random;
+    }
+
+    public SecurityProvider getSecurityProvider() {
+        return securityProvider;
+    }
+
+    public Set<SMB2Dialect> getSupportedDialects() {
+        return EnumSet.copyOf(dialects);
+    }
+
+    public UUID getClientGuid() {
+        return clientGuid;
+    }
+
+    public List<Factory.Named<Authenticator>> getSupportedAuthenticators() {
+        return new ArrayList<>(authenticators);
+    }
+
+    public boolean isSigningRequired() {
+        return signingRequired;
+    }
+
+    public int getReadBufferSize() {
+        return readBufferSize;
+    }
+
+    public int getWriteBufferSize() {
+        return writeBufferSize;
+    }
+
+    public int getTransactBufferSize() {
+        return transactBufferSize;
+    }
+
+    public static class Builder {
+        private Config config;
+
+        Builder() {
+            config = new Config();
+            setClientGuid(UUID.randomUUID());
+            setRandomProvider(new SecureRandom());
+            setSecurityProvider(new JceSecurityProvider());
+            setReadBufferSize(SMB2Packet.SINGLE_CREDIT_PAYLOAD_SIZE);
```

Comment:
Maybe have the default be `Integer.MAX_INT`, which means that they'll be taken from the negotiated protocol. Now they'll always be taken from the config, unless you manually set it to MAX_INT (or something higher than negotiated). I think the default should be that it is taken from the negotiation, and can be tweaked from the config.